### PR TITLE
Add PostgresDynamicTypeThrowingEncodable and PostgresDynamicTypeEncodable

### LINKED ISSUE #333: Support Types that have a dynamic Postgres type
The current PostgresEncodable:

```swift
/// A type that can encode itself to a postgres wire binary representation.
public protocol PostgresEncodable {
    // TODO: Rename to `PostgresThrowingEncodable` with next major release

    /// identifies the data type that we will encode into `byteBuffer` in `encode`
    static var psqlType: PostgresDataType { get }

    /// identifies the postgres format that is used to encode the value into `byteBuffer` in `encode`
    static var psqlFormat: PostgresFormat { get }

    /// Encode the entity into the `byteBuffer` in Postgres binary format, without setting
    /// the byte count. This method is called from the ``PostgresBindings``.
    func encode<JSONEncoder: PostgresJSONEncoder>(into byteBuffer: inout ByteBuffer, context: PostgresEncodingContext<JSONEncoder>) throws
}
```

requires that all instances of a type are encoded into the same `PostgresDataType` because of the `static` requirement. This is bad, specifically because supporting custom types (like Postgres enums) requires a dynamic OID.

For this reason @gwynne proposes adding a `PostgresDynamicTypeEncodable` and `PostgresDynamicTypeThrowingEncodable`, that use dynamic psqlType and psqlFormat. `PostgresEncodable` should then inherit from the `PostgresDynamicTypeThrowingEncodable` and `PostgresNonThrowingEncodable` from `PostgresDynamicTypeEncodable`. Love how we ended up at the place with `NonThrowing` and just `Throwing`.

### PR DESCRIPTION
**Pull request by marius-se**
Closes #333

Example:
```swift
// Create type
struct Vector: PostgresDynamicTypeNonThrowingEncodable {
    let value: [Float]

    var psqlType: PostgresDataType
    var psqlFormat: PostgresFormat

    func encode<JSONEncoder>(
        into byteBuffer: inout ByteBuffer,
        context: PostgresNIO.PostgresEncodingContext<JSONEncoder>
    ) where JSONEncoder: PostgresJSONEncoder {
        // ...
    }
}

// Create instance with psqlType and psqlFormat
let vector = Vector(value: [1, 2, 3], psqlType: 16345, psqlFormat: .binary)

// Use in query
let query: PostgresQuery = """
INSERT INTO foo (vector) SET (\(vector));
"""
```

### CODE REVIEW

File: Sources/PostgresNIO/New/PostgresCodable.swift
Comment (fabianfett) [inline Sources/PostgresNIO/New/PostgresCodable.swift:60]: Let's make this just `PostgresDynamicTypeEncodable`

File: Sources/PostgresNIO/New/PostgresQuery.swift
Comment (fabianfett) [inline Sources/PostgresNIO/New/PostgresQuery.swift:93]: We should change the above append methods and use protocol trampolines here.

File: Sources/PostgresNIO/New/PostgresQuery.swift
Comment (fabianfett) [inline Sources/PostgresNIO/New/PostgresQuery.swift:141]: Change above method and use protocol trampolines.

File: Sources/PostgresNIO/New/PostgresCodable.swift
Comment (fabianfett) [inline Sources/PostgresNIO/New/PostgresCodable.swift:26]: This should be `PostgresDynamicTypeThrowingEncodable`

File: Sources/PostgresNIO/New/PostgresCodable.swift
Comment (fabianfett) [inline Sources/PostgresNIO/New/PostgresCodable.swift:6]:
```suggestion
public protocol PostgresEncodable: PostgresDynamicTypeThrowingEncodable {
```

File: Sources/PostgresNIO/New/PostgresCodable.swift
Comment (fabianfett) [inline Sources/PostgresNIO/New/PostgresCodable.swift:60]: This should be `PostgresDynamicTypeEncodable`
```suggestion
public protocol PostgresDynamicTypeEncodable: PostgresDynamicTypeThrowingEncodable {
```

**Review by fabianfett — CHANGES_REQUESTED** [review]
Naming is a bit off right now.

File: Sources/PostgresNIO/New/PostgresCodable.swift
Comment (fabianfett) [inline Sources/PostgresNIO/New/PostgresCodable.swift:20]: This method is inherited, isn't it?

File: docker-compose.yml
Comment (fabianfett) [inline docker-compose.yml:15]: Can we do this in a separate pr please?

File: Sources/PostgresNIO/New/PostgresCodable.swift
Comment (marius-se) [reply Sources/PostgresNIO/New/PostgresCodable.swift]: For the transition phase (`// TODO: Rename to `PostgresThrowingEncodable` with next major release`)  this will be super confusing. So until the next major version `PostgresEncodable` will be the throwing version and `PostgresDynamicTypeEncodable` the non-throwing one, I'll have to put a mental note somewhere.
Tho I get the point 😆 Updated! ✅

File: Sources/PostgresNIO/New/PostgresCodable.swift
Comment (marius-se) [reply Sources/PostgresNIO/New/PostgresCodable.swift]: Yep, also removed it from `PostgresNonThrowingEncodable` since that inherits `encode` from `PostgresDynamicTypeEncodable` now ([1801353](https://github.com/vapor/postgres-nio/pull/365/commits/18013530ad476150106ae4e26e14863a888988b3))

**Comment by marius-se** [comment]
CI complains about breaking api changes, which is a bit worrying. However we have default implementations for the new non-static `psqlType` and `psqlFormat` properties, and the rest is just inheritance bubbling up from the two new protocols... sooo that shouldn't cause any problems - or does it?

File: Sources/PostgresNIO/New/PostgresCodable.swift
Comment (fabianfett) [inline Sources/PostgresNIO/New/PostgresCodable.swift:124]: this isn't needed since a `PostgresNonThrowingEncodable` also is a `PostgresEncodable`

File: Sources/PostgresNIO/New/PostgresCodable.swift
Comment (fabianfett) [inline Sources/PostgresNIO/New/PostgresCodable.swift:118]: We should ensure that the compiler can see the implementation here to allow full specialization:
```suggestion
    @inlinable
    public var psqlType: PostgresDataType { Self.psqlType }
    @inlinable
    public var psqlFormat: PostgresFormat { Self.psqlFormat }
```

File: Sources/PostgresNIO/New/PostgresCodable.swift
Comment (fabianfett) [inline Sources/PostgresNIO/New/PostgresCodable.swift:166]: We should be able to remove the implementations L126-L164 here. Those new impl. replace the old ones.

**Review by fabianfett — COMMENTED** [review]
A few more comments.

File: Sources/PostgresNIO/New/Data/Range+PostgresCodable.swift
Comment (marius-se) [reply Sources/PostgresNIO/New/Data/Range+PostgresCodable.swift:196]: https://github.com/vapor/postgres-nio/pull/365#discussion_r1189823374

File: Sources/PostgresNIO/New/Data/Range+PostgresCodable.swift
Comment (marius-se) [reply Sources/PostgresNIO/New/Data/Range+PostgresCodable.swift:235]: https://github.com/vapor/postgres-nio/pull/365#discussion_r1189823374

File: Sources/PostgresNIO/New/Data/Range+PostgresCodable.swift
Comment (marius-se) [reply Sources/PostgresNIO/New/Data/Range+PostgresCodable.swift:283]: https://github.com/vapor/postgres-nio/pull/365#discussion_r1189823374

File: Sources/PostgresNIO/New/Data/Array+PostgresCodable.swift
Comment (marius-se) [reply Sources/PostgresNIO/New/Data/Array+PostgresCodable.swift:177]: wow that's from 2019 - ancient. On it!

**Review by fabianfett — COMMENTED** [review]
I think now is the correct time to get a review from @gwynne.

File: Sources/PostgresNIO/New/Data/Range+PostgresCodable.swift
Comment (gwynne) [inline Sources/PostgresNIO/New/Data/Range+PostgresCodable.swift:196]: I really wish during review of the `PostgresRange` PR I'd thought to suggest following the stdlib convention and naming it `Bound` instead of `B` 😕

File: Sources/PostgresNIO/New/PostgresCodable.swift
Comment (gwynne) [inline Sources/PostgresNIO/New/PostgresCodable.swift:7]:
```suggestion
/// Dynamic types are types that don't have a well-known Postgres type OID at compile time.
/// For example, custom types created at runtime, such as enums, or extension types whose OID is not stable between databases.
```

File: Sources/PostgresNIO/New/PostgresCodable.swift
Comment (gwynne) [inline Sources/PostgresNIO/New/PostgresCodable.swift:45]:
```suggestion
    /// The data type encoded into the `byteBuffer` in ``encode(into:context:)``.
```

File: Sources/PostgresNIO/New/PostgresCodable.swift
Comment (gwynne) [inline Sources/PostgresNIO/New/PostgresCodable.swift:48]:
```suggestion
    /// The Postgres encoding format used to encode the value into `byteBuffer` in ``encode(into:context:)``.
```

File: Sources/PostgresNIO/New/PostgresCodable.swift
Comment (gwynne) [inline Sources/PostgresNIO/New/PostgresCodable.swift:8]:
```suggestion
public protocol PostgresThrowingDynamicTypeEncodable {
```
My reasoning for the different naming: `PostgresDynamicTypeThrowingEncodable` reads to me as "an encodable throwing a Postgres dynamic type", rather than "a throwing dynamically-typed encodable".

File: Sources/PostgresNIO/New/PostgresCodable.swift
Comment (gwynne) [inline Sources/PostgresNIO/New/PostgresCodable.swift:9]:
```suggestion
    /// The data type encoded into the `byteBuffer` in ``encode(into:context:)``.
```

**Review by gwynne — CHANGES_REQUESTED** [review]
A pile of nits, several about wording of comments, a couple about naming.

An additional thought: Would it make sense to also provide utility methods for looking up type OIDs by name (e.g. wrapping `SELECT 'foo'::"regtype"::"integer" as "oid"`), or do we want to leave that to the purview of higher-level packages?

File: Sources/PostgresNIO/New/Data/Range+PostgresCodable.swift
Comment (gwynne) [reply Sources/PostgresNIO/New/Data/Range+PostgresCodable.swift:196]: Well then let's fix it! 😛

Comment (fabianfett) [reply Sources/PostgresNIO/New/Data/Range+PostgresCodable.swift:196]: and this is internal anyway. fixing.

Comment (fabianfett) [reply Sources/PostgresNIO/New/Data/Range+PostgresCodable.swift:196]: Fixed and landed #367, sorry for the merge conflict @marius-se

**Comment by marius-se** [comment]
> A pile of nits, several about wording of comments, a couple about naming.
> 
> An additional thought: Would it make sense to also provide utility methods for looking up type OIDs by name (e.g. wrapping `SELECT 'foo'::"regtype"::"integer" as "oid"`), or do we want to leave that to the purview of higher-level packages?

Hmmm personally I feel like this belongs into a higher-level package. As far as I understand PostgresNIO is supposed to provide an interface to build and execute queries, not offer a toolbox of ready-to-use queries. What do you think?

**Comment by gwynne** [comment]
> Hmmm personally I feel like this belongs into a higher-level package. As far as I understand PostgresNIO is supposed to provide an interface to build and execute queries, not offer a toolbox of ready-to-use queries. What do you think?

I tend to agree; I pretty much only brought it up to make sure we're all on the same page 🙂.

**Review by gwynne — APPROVED** [review]
Ship it! 😂 

This Is wonderful work! ❤️

**Comment by marius-se** [comment]
@gwynne What about [this](https://github.com/vapor/postgres-nio/pull/365#discussion_r1190118932) comment?
> ...providing func encode<S>(into byteBuffer: inout ByteBuffer, context: PostgresEncodingContext<PostgresJSONEncoder>) where S: Sequence, S.Element: PostgresArrayEncodable { ... } instead of trying to add protocol conformance.

We would still need to add protocol conformance to the different sequence types, wouldn't we?

**Comment by marius-se** [comment]
Good to hear that you like it! 😁

**Comment by gwynne** [comment]
> We would still need to add protocol conformance to the different sequence types, wouldn't we?

Exactly the opposite - withholding the conformance from any sequence types (including `Array`) would remove any ambiguity about which version of `encode()` to call, although my original comment should have specified that the additional method(s) would have to be available on `PostgresBindings`, not directly on the various encodable protocols. (That's actually what makes it a no more ideal solution than the suggestion of conforming the sequence types one by one - the ability to invoke the encoding logic independently of any supporting types is lost, just as it was when working with `any Encodable` pre-Swift 5.7.)

**Comment by marius-se** [comment]
Ahhh got it!

**Comment by gwynne** [comment]
@fabianfett Ping on this - do you still have any blocking concerns?

@marius-se Ping as well - can you resolve the merge conflicts and bring this up to date with `main`?

**Comment by fabianfett** [comment]
IIRC all good from my side. @marius-se can you bring this in line with `main`? Once done I'll do a final review. Thanks for your work up until here.

**Comment by fabianfett** [comment]
@marius-se small ping... shall I move this forward?

**Comment by marius-se** [comment]
Oh sorry @fabianfett! I'm currently traveling through New Zealand (without electricity 😅). If you need the changes on main now feel free to take this over! Otherwise I'll be back in 3-4 weeks :)

**Comment by fabianfett** [comment]
API breakage checker complain. Will need to research this one:
```
21 breaking changes detected in PostgresNIO:
  💔 API breakage: protocol PostgresArrayEncodable has added inherited protocol PostgresThrowingDynamicTypeEncodable
  💔 API breakage: protocol PostgresRangeEncodable has added inherited protocol PostgresDynamicTypeEncodable
  💔 API breakage: protocol PostgresRangeEncodable has added inherited protocol PostgresThrowingDynamicTypeEncodable
  💔 API breakage: protocol PostgresRangeArrayEncodable has added inherited protocol PostgresDynamicTypeEncodable
  💔 API breakage: protocol PostgresRangeArrayEncodable has added inherited protocol PostgresThrowingDynamicTypeEncodable
  💔 API breakage: protocol PostgresEncodable has generic signature change from  to <Self : PostgresNIO.PostgresThrowingDynamicTypeEncodable>
  💔 API breakage: protocol PostgresEncodable has added inherited protocol PostgresThrowingDynamicTypeEncodable
  💔 API breakage: protocol PostgresNonThrowingEncodable has generic signature change from <Self : PostgresNIO.PostgresEncodable> to <Self : PostgresNIO.PostgresDynamicTypeEncodable, Self : PostgresNIO.PostgresEncodable>
  💔 API breakage: protocol PostgresNonThrowingEncodable has added inherited protocol PostgresDynamicTypeEncodable
  💔 API breakage: protocol PostgresNonThrowingEncodable has added inherited protocol PostgresThrowingDynamicTypeEncodable
  💔 API breakage: func PostgresQuery.StringInterpolation.appendInterpolation(_:) has generic signature change from <Value where Value : PostgresNIO.PostgresEncodable> to <Value where Value : PostgresNIO.PostgresThrowingDynamicTypeEncodable>
  💔 API breakage: func PostgresQuery.StringInterpolation.appendInterpolation(_:) has generic signature change from <Value where Value : PostgresNIO.PostgresEncodable> to <Value where Value : PostgresNIO.PostgresThrowingDynamicTypeEncodable>
  💔 API breakage: func PostgresQuery.StringInterpolation.appendInterpolation(_:) has generic signature change from <Value where Value : PostgresNIO.PostgresNonThrowingEncodable> to <Value where Value : PostgresNIO.PostgresDynamicTypeEncodable>
  💔 API breakage: func PostgresQuery.StringInterpolation.appendInterpolation(_:) has generic signature change from <Value where Value : PostgresNIO.PostgresNonThrowingEncodable> to <Value where Value : PostgresNIO.PostgresDynamicTypeEncodable>
  💔 API breakage: func PostgresQuery.StringInterpolation.appendInterpolation(_:context:) has generic signature change from <Value, JSONEncoder where Value : PostgresNIO.PostgresEncodable, JSONEncoder : PostgresNIO.PostgresJSONEncoder> to <Value, JSONEncoder where Value : PostgresNIO.PostgresThrowingDynamicTypeEncodable, JSONEncoder : PostgresNIO.PostgresJSONEncoder>
  💔 API breakage: func PostgresBindings.append(_:) has generic signature change from <Value where Value : PostgresNIO.PostgresEncodable> to <Value where Value : PostgresNIO.PostgresThrowingDynamicTypeEncodable>
  💔 API breakage: func PostgresBindings.append(_:context:) has generic signature change from <Value, JSONEncoder where Value : PostgresNIO.PostgresEncodable, JSONEncoder : PostgresNIO.PostgresJSONEncoder> to <Value, JSONEncoder where Value : PostgresNIO.PostgresThrowingDynamicTypeEncodable, JSONEncoder : PostgresNIO.PostgresJSONEncoder>
  💔 API breakage: func PostgresBindings.append(_:) has generic signature change from <Value where Value : PostgresNIO.PostgresNonThrowingEncodable> to <Value where Value : PostgresNIO.PostgresDynamicTypeEncodable>
  💔 API breakage: func PostgresBindings.append(_:context:) has generic signature change from <Value, JSONEncoder where Value : PostgresNIO.PostgresNonThrowingEncodable, JSONEncoder : PostgresNIO.PostgresJSONEncoder> to <Value, JSONEncoder where Value : PostgresNIO.PostgresDynamicTypeEncodable, JSONEncoder : PostgresNIO.PostgresJSONEncoder>
  💔 API breakage: func PostgresEncodable.encode(into:context:) has been removed
  💔 API breakage: func PostgresNonThrowingEncodable.encode(into:context:) has been removed
```

**Review by fabianfett — APPROVED** [review]
LGTM. The API breakage checker complains are false positives.

**Comment by fabianfett** [comment]
@marius-se Thanks so much for pushing this through!

**Pull request merged**
